docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,115 @@
+  ability to introspect the implementations of reactors in most
+  cases.
+
+- the exception to this, is the cluster of
+  consensus/statesync/blockhain (which depend on eachother,) because
+  `node` itself contains some of their specific initialization logic.
+
+- while the RPC environment construction happens in `node`, this can
+  happen early during `node` construction rather than at
+  initialization time, which reduces the need for the implementation
+  of the node object to track this state.
+
+Ideally, it seems like a node instance should be expressable simply as
+a list of `service.Services` and any knowledge of the internals of any
+of these services could be limited to their constructors. I sketched
+out a generic and simple `groupService` implementation, to handle a
+desecrate stage or sequence of services.  As some services have
+dependencies during construction or initialization and as a starting
+point, a cluster of dependencies can be wrapped as a single
+`service.Service`.
+
+## Goals
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects. 
+  
+  - Reactors should not need to expose their interfaces *within* the
+    implementation of the node type, except in the context of some groups of
+    service. 
+
+- Expose some "unsafe"  way of replacing (some?) existing components, and or
+  inserting arbitrary reactors or services into the tendermint node's
+  initialization.
+  
+  - This may include creating an exported, top-level "unsafe" package with
+    several interfaces and constructors, or simply a function in the
+    node package. 
+
+- Refactor and simplify the process start/initialization logic,
+  separately from--but as a precursor to--permitting user replacement
+  of components.
+
+## Non-Goals
+
+- Fully abstract dependency-based process initialization. 
+
+- Fully abstract replacement of core tendermint components without vendoring.
+
+- Parallelized initialization of components.
+
+## Questions
+
+- To what extent does this new initialization framework need to accommodate
+  the legacy p2p stack? Would it be possible to delay a great deal of this
+  work to the 0.36 cycle to avoid this complexity? 
+  
+- There's clear and common interest in injecting some kind of mempool, and
+  intuitively this makes sense as the highest priority.
+  
+  - Does the existing `mempool.Mempool` interface make sense to expose and
+    commit to, or should we make a smaller interface? 
+    
+  - Is it reasonable for users to bring their own mempool without bringing
+    their own reactor?
+
+- Are there reactors, other than the mempool, that make sense to support
+  ad-hoc replacing (e.g. statesync, blockchain (fastsync),) and if so should
+  we provide any first-class support for this?
+
+- There's a dependency (and nearly a cycle), between consensus,
```

Comment:
until or if we refactor consensus we should not allow to exchange it because it is so tightly woven into the repo it could cause unknown issues.

Reply:
Yeah, I think this is in general the argument for not implementing or providing plugability at this juncture.

Reply:
I would say this mainly applies to consensus. The mempool could be replaced with little to no issue.